File: offer/ex25/hasPath.go
// Copyright 2019 readailib.com. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package ex25

/*
# -*- coding:utf-8 -*-
class Solution:
    def hasPath(self, matrix, rows, cols, path):
        # write code here
*/

// 矩阵中的路径
func hasPath(matrix []string, rows int, cols int, path string) bool {
	// 这个是标记数组，用来标记当前是否访问过
	flag := make([]bool, len(matrix))

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if judge(matrix, rows, cols, flag, i, j, path, 0) {
				return true
			}
		}
	}
	return false
}

// level as the current charactor
func judge(matrix []string, rows int, cols int, flag []bool, i int, j int, path string, level int) bool {

	// 通过i，j计算当前的 索引位置
	index := i*cols + j

	// 判断是否满足递归的终止条件
	if i < 0 || j < 0 || i >= rows || j >= cols || matrix[index] != string(path[level]) || flag[index] == true {
		return false
	}

	// 标记当前的已被访问
	flag[index] = true

	if judge(matrix, rows, cols, flag, i+1, j, path, level+1) ||
		judge(matrix, rows, cols, flag, i-1, j, path, level+1) ||
		judge(matrix, rows, cols, flag, i, j+1, path, level+1) ||
		judge(matrix, rows, cols, flag, i, j-1, path, level+1) {
		return true
	}

	// 如果不满足上面的条件，将上面的标记为false
	flag[index] = false
	return false
}

File: offer/ex25/hasPath.py
# -*- coding:utf-8 -*-


# 请设计一个函数，用来判断在一个矩阵中是否存在一条包含某字符串所有字符的路径。
# 路径可以从矩阵中的任意一个格子开始，
# 每一步可以在矩阵中向左，向右，向上，向下移动一个格子。
# 如果一条路径经过了矩阵中的某一个格子，则该路径不能再进入该格子。 
# 
# 例如 
#      a b c e 
#      s f c s 
#      a d e e 
# 矩阵中包含一条字符串"bcced"的路径，
# 但是矩阵中不包含"abcb"路径，
# 因为字符串的第一个字符b占据了矩阵中的第一行第二个格子之后，
# 路径不能再次进入该格子。

# -*- coding:utf-8 -*-
class Solution:
    
    def hasPath(self, matrix, rows, cols, path):
        # write code here
        flag = [False for _ in range(len(matrix))]
        for i in range(rows):
            for j in range(cols):
                # 遍历
                if self.judge(matrix, rows, cols, flag, i, j, path, 0):
                    return True
        return False

    def judge(self, matrix, rows, cols, flag, i, j, path, level):
        index = i * cols + j
        # 判断越界条件,递归终止条件
        if i >= rows or j >= cols or i < 0 or j < 0 or \
                matrix[index] != path[level] or \
                flag[index] == True:
            return False

        if level == len(path) - 1:
            return True

        flag[index] = True

        if (self.judge(matrix, rows, cols, flag, i + 1, j, path, level + 1) or \
                self.judge(matrix, rows, cols, flag, i, j + 1, path, level + 1) or \
                self.judge(matrix, rows, cols, flag, i - 1, j, path, level + 1) or \
                self.judge(matrix, rows, cols, flag, i, j - 1, path, level + 1)):
            return True

        flag[index] = False
        return False
